### ISSUE TITLE
Use Two Remotes Control with BrickPI3

### ISSUE THREAD

**Issue by Negociation** [ev3dev/ev3dev#1176]
### What are you trying to do?
Hi , i posted before here , i trying to do a self driving robot using ev3dev + CNN , now i'm working on dataset algortym generator... 

### What did you expect to happen?
- Because the hardware recently i move from EV3 to Raspberry PI + BrickPI , i'm using Python , now to control my robot i'm using two remote controls , the first on channel 1 to forward and backward , and a second on channel 2 to left and right 

### What actually happened?
My robot only responses to my Control on Channel 1 , why thats happens ? 

My Code:
```python
#!/usr/bin/env python3
from time import sleep
import time
import os.path
from ev3dev.brickpi import *
import threading
import subprocess
import cv2
###################################################################
#		   DECLARACOES DE INICIALIZACAO			   #
###################################################################

#Dados de Revisao
Release = '3'
Version = '1'
Subversion = '2'
Last_State = 'Stop'
StopCapture = False


###########################################################################
#		   	            DATASET  			          #
###########################################################################



class Source:
	def __init__(self):
		#Conjunto de Dados de Objetos Row
		self.dataset = []
		self.target = 'Version = '+ Release + '.' + Version + '.' + Subversion
		self.date = time.strftime("%d%m%Y")
		self.hour = time.strftime("%H:%M:%S")
		self.build = 1
		while os.path.isfile('/home/robot/r'+self.date+'-b'+str(self.build)+'.marcset'):
			self.build += 1
		self.filename = '/home/robot/r'+self.date+'-b'+str(self.build)+'.marcset';
		
	
	def getTime(self):
		return self.date + ' ' + self.hour;
		

	def getBuild(self):
		return self.build
	
	def Insert(self,Row):
		self.dataset.append(Row)
	
	def getSize(self):
		return len(self.dataset)
	
	def getRow(self,Index):
		return self.dataset[Index]

	def getDataset(self):
		return self.dataset

	def getFilename(self):
		return self.filename
	
Dataset = Source()	

class Row:
	def __init__(self,RowTempo,RowLuz,RowProximidade,RowClasse):
		self.rowTempo = RowTempo
		self.rowLuz = RowLuz
		self.rowProximidade = RowProximidade 
		self.rowClasse = RowClasse

	def get_RowTempo(self):
		return self.rowTempo

	def get_RowLuz(self):
		return self.rowLuz

	def get_RowProximidade(self):
		return self.rowProximidade

	def get_RowClasse(self):
		return self.rowClasse

		


###########################################################################
#		   	          MEU ROBO       		          #
###########################################################################

class Robot:
	def __init__(self):
		self.robot_MotorDireito = None
		self.robot_MotorEsquerdo = None
		self.robot_SensorUltrasonico = None
		self.robot_SensorInfravermelho = None

	#Setando Porta do Motor Direito
	def set_MotorDireito(self,motorDireito):
		self.robot_MotorDireito = LargeMotor(motorDireito)
	
	#Setando Porta do Motor Esquerdo
	def set_MotorEsquerdo(self,motorEsquerdo):
		self.robot_MotorEsquerdo = LargeMotor(motorEsquerdo)

	#Setando Porta do Sensor Ultrasonico
	def set_SensorUltrasonico(self,sensorUltrasonico):
		self.robot_SensorUltrasonico = UltrasonicSensor(sensorUltrasonico)

	#Setando Modo do Sensor Ultrasonico
	def set_SensorUltrasonicoMode(self,sensorUltrasonicoMode):
		self.robot_SensorUltrasonico.mode = sensorUltrasonicoMode

	#Capturando Dados de Entrada do Sensor de Ultrasonico
	def get_SensorUltrasonico(self):
		return self.robot_SensorUltrasonico.value()/10

	#Setando Porta do Sensor Infravermelho
	def set_SensorInfravermelho(self,sensorInfravermelho):
		self.robot_SensorInfravermelho = InfraredSensor(sensorInfravermelho)

	#Setando Modo do Sensor Infravermelho
	def set_SensorInfravermelhoMode(self,sensorInfravermelhoMode):
		self.robot_SensorInfravermelho.mode = sensorInfravermelhoMode

	#Setando Porta do Sensor de Cor/Luminosidade
	def set_SensorLuminosidade(self,sensorLuminosidade):
		self.robot_SensorLuminosidade = ColorSensor(sensorLuminosidade)

	#Setando Modo do Sensor de Cor/Luminosidade
	def set_SensorLuminosidadeMode(self,sensorLuminosidadeMode):
		self.robot_SensorLuminosidade.mode = sensorLuminosidadeMode

	#Capturando Dados de Entrada do Sensor de Luminosidade
	def get_SensorLuminosidade(self):
		return self.robot_SensorLuminosidade.value()
	
	#Validacao de Movimentos do Robo
	def Validate(self,movimento):
		if movimento=="Foward":
			self.robot_MotorDireito.run_forever(speed_sp=-500)
			self.robot_MotorEsquerdo.run_forever(speed_sp=-500)
		elif movimento=="Stop":
			self.robot_MotorDireito.stop(stop_action='coast')
			self.robot_MotorEsquerdo.stop(stop_action='coast')
			myRobot.Instance('Stop');
		elif movimento=="Left":
			self.robot_MotorDireito.run_forever(speed_sp=100)	
			self.robot_MotorEsquerdo.run_forever(speed_sp=-200)
		elif movimento=="Right":
			self.robot_MotorEsquerdo.run_forever(speed_sp=100)	
			self.robot_MotorDireito.run_forever(speed_sp=-200)

	def Instance(self,Movimento):
		myRow = Row('0:00:00',self.get_SensorLuminosidade(),self.get_SensorUltrasonico(),Movimento)
		Dataset.Insert(myRow)
	
	def Stop(self):
		self.robot_MotorDireito.stop(stop_action="coast");
		self.robot_MotorEsquerdo.stop(stop_action="coast");


#Instanciando meu Robo
myRobot = Robot() 

#Motores
myRobot.set_MotorDireito('MB')
myRobot.set_MotorEsquerdo('MC')

#Sensores
myRobot.set_SensorUltrasonico('S1')
myRobot.set_SensorUltrasonicoMode('US-DIST-CM')
myRobot.set_SensorInfravermelho('S4')
myRobot.set_SensorInfravermelhoMode('IR-REMOTE')
myRobot.set_SensorLuminosidade('S3')
myRobot.set_SensorLuminosidadeMode('COL-AMBIENT')


#Declaracao do Controle Remoto Esquerdo no Canal 1
ListenA = RemoteControl(channel=1)
ActionA = RemoteControl(channel=1)


#Declaracao do Controle Remoto Direito no Canal 4
ListenB = RemoteControl(channel=2)
ActionB = RemoteControl(channel=2)



###########################################################################
#		   	         EVENTOS LISTEN			         #
###########################################################################

#Funcao Referentes ao [ LISTEN ] do Controle Remoto A [ LEFT ]
def Listen_Left():
	def on_press(state):
		global Last_State
		if (state):
			Last_State = 'Left'
			#Leds.all_off()
			#Leds.set_color(Leds.LEFT,  Leds.GREEN)
	return on_press



#Funcao Referentes ao [ LISTEN ] do Controle Remoto A [ RIGHT ]
def Listen_Right():
	def on_press(state):
		global Last_State
		if (state):
			Last_State = 'Right'
			#Leds.all_off()
			#Leds.set_color(Leds.RIGHT,  Leds.GREEN)
	return on_press



#Funcao Referentes ao [ LISTEN ] do Controle Remoto B [ TOP ]
def Listen_Top():
	def on_press(state):
		global Last_State
		if (state):
			Last_State = "Foward"
			#Leds.all_off()
			#Leds.set_color(Leds.LEFT,  Leds.GREEN)
			#Leds.set_color(Leds.RIGHT,  Leds.GREEN)
	return on_press



#Funcao Referentes ao [ LISTEN ] do Controle Remoto A [ DOWN ]
def Listen_Down():
	def on_press(state):
		print("algo sim")
		global Last_State
		global StopCapture
		if (state):
			Last_State = "Stop"
			#Leds.all_off()
			#Leds.set_color(Leds.LEFT,  Leds.RED)
			#Leds.set_color(Leds.RIGHT,  Leds.RED)
			StopCapture = True
	return on_press


#Atribuicao de Funcoes ao Evento onPressed do Controle Remoto A [ LISTEN ]
ListenA.on_red_up = Listen_Left()
ListenA.on_red_down = Listen_Left()
ListenA.on_blue_up = Listen_Right()
ListenA.on_blue_down = Listen_Right()



#Atribuicao de Funcoes ao Evento onPressed do Controle Remoto B [ LISTEN ]
ListenB.on_red_up = Listen_Top()
ListenB.on_blue_up = Listen_Top()
ListenB.on_red_down = Listen_Down()
ListenB.on_blue_down = Listen_Down()




###########################################################################
#		   	         EVENTOS ACTION			          #
###########################################################################


#Funcao Referentes ao [ ACTION ] do Controle Remoto A [ LEFT ]
def Action_Left():
	def on_press(state):
		if (state):
			myRobot.Validate('Left')
	return on_press



#Funcao Referentes ao [ ACTION ] do Controle Remoto A [ RIGHT ]
def Action_Right():
	def on_press(state):
		if (state):
			myRobot.Validate('Right')
	return on_press



#Funcao Referentes ao [ ACTION ] do Controle Remoto B [ TOP ]
def Action_Top():
	def on_press(state):
		if (state):
			myRobot.Validate('Foward')
	return on_press



#Funcao Referentes ao [ ACTION ] do Controle Remoto A [ DOWN ]
def Action_Down():
	def on_press(state):
		if (state):
			myRobot.Validate('Stop') 
	return on_press



#Atribuicao de Funcoes ao Evento onPressed do Controle Remoto A [ ACTION ]
ActionA.on_red_up = Action_Left()
ActionA.on_red_down = Action_Left()
ActionA.on_blue_up = Action_Right()
ActionA.on_blue_down = Action_Right()



#Atribuicao de Funcoes ao Evento onPressed do Controle Remoto B [ ACTION ]
ActionB.on_red_up = Action_Top()
ActionB.on_blue_up = Action_Top()
ActionB.on_red_down = Action_Down()
ActionB.on_blue_down = Action_Down()


#Thread Para Leitura dos Sensores com base nos movimentos Controle Remoto
class Thread_Input (threading.Thread):
	global Last_State
	global StopCapture
	def __init__(self):
		threading.Thread.__init__(self)
	def run(self):
		while (StopCapture == False):	
			if(Last_State != 'Stop' ):
				myRobot.Instance(Last_State)			
				sleep(1)
		myRobot.Stop();

#Thread Para Movimento com uso dos Controles Remotos
class Thread_Output (threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)
	def run(self):
		global StopCapture
		while (StopCapture == False):	
			ActionA.process()
			sleep(0.01)
			ActionB.process()
			sleep(0.01)


#Thread Para Set de Movimentos e Led
class Thread_Setting (threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)
	def run(self):
		global StopCapture
		while(StopCapture == False):	
			ListenA.process()
			sleep(0.01)
			ListenB.process()

#Inicializando Sensores nas Portas ( BrickPI )
#subprocess.call("./Set_Device.sh",shell=True)


#Inicializacao da Thread Input( Movimento )
Input = Thread_Input()
Input.start()

#Inicializacao da Thread Input( Movimento )
Setting = Thread_Setting()
Setting.start()


#Inicializacao da Thread Output ( Movimento )
Output = Thread_Output()
Output.start()

#Mensagem de Boas Vindas
print('\n\n\n')
VersionData = Release + '.' + Version + '.' +Subversion
print('Inicializando o Sistema de Captura de Dados (SCD)',VersionData)
print('----------------------------------------------------------')

#Bip de Inicializacao
Sound.beep()

def WriteFile():
	MARCset = open(Dataset.getFilename(), 'w+')

	Header = "------------------------------------\n!    Sistema de Captura de Dados   !\n------------------------------------\n! tiagitofranca04@gmail.com        !\n------------------------------------\n"
	SubHeader = '+ Versao da Compilacao < '+str(Release) + '.' + str(Version) + '.' +str(Subversion)+' >\n+ Data da Captura <'+str(Dataset.getTime())+' >\n+ Build < '+ str(Dataset.getBuild()) +' >\n\nMARCSET:\n'
	MARCset.writelines(Header)
	MARCset.writelines(SubHeader)
	for Row in Dataset.getDataset(): 
		Datarow = str(Row.get_RowTempo()) + " " + str(Row.get_RowProximidade()) + " " + str(Row.get_RowLuz()) + " " + str(Row.get_RowClasse()) + "\n"
		MARCset.writelines(Datarow)
	MARCset.close()


while(StopCapture == False):
	LastCount = 0
	sleep(1)
	if (Dataset.getSize() != LastCount):
		LastCount = Dataset.getSize()
		myObject = Dataset.getRow(Dataset.getSize()-1);
		print(str(Dataset.getSize()) + ' >> T: ' + str(myObject.get_RowTempo()) + ' - D: ' + str(myObject.get_RowProximidade()) + ' - L: ' + str(myObject.get_RowLuz()) + ' - C: ' + str(myObject.get_RowClasse()))
	sleep(0.1)

WriteFile()


```

**Comment by dlech** [comment]
You missed an important question:

```
### What hardware and software are you using (including version numbers)?

<!-- Tip: copy and paste the output of `ev3dev-sysinfo -m` here -->
```

**Comment by dlech** [comment]
cc: @ev3dev/python

**Comment by Negociation** [comment]
Image file:         ev3dev-stretch-rpi2-generic-2018-08-06
Kernel version:     4.14.59-ev3dev-2.2.2-rpi2
Brickman:           0.10.0
BogoMIPS:           38.40
Board:              board0
BOARD_INFO_HW_REV=A52082
BOARD_INFO_MODEL=Raspberry Pi 3 Model B Rev 1.2
BOARD_INFO_SERIAL_NUM=000000005b74a978
BOARD_INFO_TYPE=main
Board:              board1
BOARD_INFO_FW_VER=1.4.3
BOARD_INFO_HW_REV=3.2.1
BOARD_INFO_MODEL=Dexter Industries BrickPi3
BOARD_INFO_SERIAL_NUM=46E429C9515035524E4A2020FF071B2B
BOARD_INFO_TYPE=aux
```

**Comment by dlech** [comment]
One thing I noticed is that since you are using BrickPi3 you should be using:

```python
from ev3dev.brickpi3 import *
```

not 

```python
from ev3dev.brickpi import *
```

**Comment by Negociation** [comment]
Oh yeah , about that , its because i couldnt find a good documentation or examples of ev3python for BrickPI3 here , i loses 2 days trying to set the ports correctly huahuhauha... i probably tried change the import... 

EDIT: Still not working :/

- I Tested change the channels of functions , apparently only works with channel 1 , i don't know why ...

EDIT2: I tested on watch values and only channel 1 is recieving the information...

**Comment by dlech** [comment]
If this same code worked on the EV3, then I'm guessing this is a kernel driver bug in the BrickPi3 drivers.

https://github.com/ev3dev/lego-linux-drivers/blob/126c76d23443acab0a08b8069960e5308f9f4567/brickpi3/brickpi3_ports_in.c#L199

Looks like `BRICKPI3_SENSOR_TYPE_EV3_INFRARED_REMOTE` from line above should be moved to the group that reads 4 data values in line below.

https://github.com/ev3dev/lego-linux-drivers/blob/126c76d23443acab0a08b8069960e5308f9f4567/brickpi3/brickpi3_ports_in.c#L223

**Comment by Negociation** [comment]
Ok , how i do that ? , how can i edit the brickpi3_pors_in.c ? , i really want to use the EV3DEV , and if my project work my college will open a robotic course using EV3DEV on the next year based on my results :D

**Comment by dlech** [comment]
Try this kernel:

[linux-image-4.14.71-ev3dev-2.3.0-rpi2+_1_armhf.zip](https://github.com/ev3dev/ev3dev/files/2431108/linux-image-4.14.71-ev3dev-2.3.0-rpi2%2B_1_armhf.zip)

* Download the file
* Unzip it
* copy the .deb to the RPi
* Run `sudo dpkg -i linux-image-4.14.71-ev3dev-2.3.0-rpi2+_1_armhf.deb` on the RPi
* Reboot the RPi

**Comment by Negociation** [comment]
Tanks bro , i'll try now... , I imagine as something about driver , because i read when u started the BrickPI3 Support u basicly had to rewrite the drivers again from scrath , i guess i'm probably the first who trying to use two controls at the same time the reason why i doing that its because as more ergonomic, take a look: 


           LEFT CONTROL                         RIGHT CONTROL
Green: Forward , Red: Stop |  Red Left: Left , Blue Right: Right
![controle](https://user-images.githubusercontent.com/32465944/46251738-11bfee00-c432-11e8-884a-d0afe8400a4a.jpg)

PS:. I tested and not work yet , still only reading the channel 1...

**Comment by dlech** [comment]
Run `uname -r` to see if the new kernel is really being used

**Comment by dlech** [comment]
And if the picture is recent, one remote looks like it is on channel 4, but the program uses channels 1 and 2.

I like the red/green button mod, by the way. :smile:

**Comment by Negociation** [comment]
I changed again to channel 4 on my code before , and i tested on watch values but nothing... huahauha , 

4.14.71-ev3dev-2.3.0-rpi2+ and yeah using the new kernel :)

**Comment by dlech** [comment]
OK, tomorrow I will get my BrickPi3 out and try again.

**Comment by Negociation** [comment]
Tanks for helping me bro... anyway tanks for put the support for OpenCV on EV3DEV image , but its too late for me now huahua i'm using now a smartphone to recorder video...

**Comment by dlech** [comment]
[linux-image-4.14.71-ev3dev-2.3.0-rpi2+_1_armhf.zip](https://github.com/ev3dev/ev3dev/files/2431907/linux-image-4.14.71-ev3dev-2.3.0-rpi2%2B_1_armhf.zip)

Please try this kernel. All 4 channels should be working now.

**Comment by Negociation** [comment]
Tanks , i was waiting here huahuhauha, so whats happened ?

**Comment by Negociation** [comment]
Working fine now :D

**Comment by dlech** [comment]
> so whats happened ?

There were two bugs that needed to be fixed, not just one. You can look at the commit linked above if you want to know the technical details.

**Comment by Negociation** [comment]
Next Step for me now  its use some PID with Gyro Sensor...  i dont know why but my robot doesnt run straight

**Comment by dlech** [comment]
Fix released in kernel `ev3dev-2.3.1`.
